warp/Network/Wai/Handler/Warp/WithApplication.hs

```diff
@@ -0,0 +1,94 @@
+
+module Network.Wai.Handler.Warp.WithApplication (
+  withApplication,
+  testWithApplication,
+  openFreePort,
+) where
+
+import           Control.Concurrent
+import           Control.Exception
+import           Network.Socket
+import           Network.Wai
+import           Network.Wai.Handler.Warp.Run
+import           Network.Wai.Handler.Warp.Settings
+import           Network.Wai.Handler.Warp.Types
+
+data App
+  = App {
+    appThread :: ThreadId,
+    appKilled :: Waiter (),
+    appPort :: Int
+  }
+
+-- | Runs the given 'Application' on a free port. Passes the port to the given
+-- operation and executes it, while the 'Application' is running. Shuts down the
+-- server before returning.
+withApplication :: IO Application -> (Port -> IO a) -> IO a
+withApplication mkApp action = do
+  app <- mkApp
+  bracket (acquire app) free (\ runningApp -> action (appPort runningApp))
+  where
+    acquire :: Application -> IO App
+    acquire app = do
+      start <- mkWaiter
+      killed <- mkWaiter
+      thread <- forkIO $ do
+        (port, sock) <- openFreePort
+        let settings =
+              defaultSettings{
+                settingsBeforeMainLoop = notify start port
+              }
+        runSettingsSocket settings sock app
+          `finally` notify killed ()
```

Comment:
There's a slight async exception hole here (besides the `openFreePort` exception I mentioned above). How about using `forkFinally` instead, e.g.:

```haskell
(port, sock) <- openFreePort
let settings = defaultSettings {...}
thread <- forkFinally
  (runSettingsSocket settings sock app)
  (const (notify killed ())
```